bodo/pandas/physical/join.h

```diff
@@ -92,16 +92,53 @@ class PhysicalJoin : public PhysicalSourceSink, public PhysicalSink {
         std::shared_ptr<bodo::Schema> probe_table_schema_reordered =
             probe_table_schema->Project(probe_col_inds);
 
+        for (const duckdb::JoinCondition& cond : conditions) {
+            auto& left_bce =
+                cond.left->Cast<duckdb::BoundColumnRefExpression>();
+            auto& right_bce =
+                cond.right->Cast<duckdb::BoundColumnRefExpression>();
+
+            std::shared_ptr<PhysicalExpression> new_phys_expr =
+                std::static_pointer_cast<
+                    PhysicalExpression>(std::make_shared<
+                                        PhysicalComparisonExpression>(
+                    std::make_shared<PhysicalColumnRefExpression>(
+                        this->probe_col_inds_rev[getColRefMap(
+                            left_bindings)[{left_bce.binding.table_index,
+                                            left_bce.binding.column_index}]],
+                        left_bce.GetName(), true),
+                    std::make_shared<PhysicalColumnRefExpression>(
+                        this->build_col_inds_rev[getColRefMap(
+                            right_bindings)[{right_bce.binding.table_index,
+                                             right_bce.binding.column_index}]],
+                        right_bce.GetName(), false),
+                    cond.comparison));
+            if (physExprTree) {
+                physExprTree = std::static_pointer_cast<PhysicalExpression>(
+                    std::make_shared<PhysicalConjunctionExpression>(
+                        physExprTree, new_phys_expr,
+                        duckdb::ExpressionType::CONJUNCTION_AND));
+            } else {
+                physExprTree = new_phys_expr;
+            }
+        }
+
+        // ---------------------------------------------------
+
         // TODO[BSE-4813]: handle outer joins properly
         bool build_table_outer = false;
         bool probe_table_outer = false;
 
+        cond_expr_fn_t join_func = nullptr;
+        if (has_non_equi_cond) {
+            join_func = PhysicalExpression::join_expr;
```

Comment:
Calling the expression evaluator on each row can be expensive. Let's open an issue for performance optimization. Ideally, we should JIT the expression and pass in batches of rows. Let's see how TPC-H queries using this perform and decide on priority.